File: solutions/c/dnd-character/1/dnd_character.c
#include <stdlib.h>
#include <time.h>
#include "dnd_character.h"

static int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

int ability(void) {
    int rolls[4];
    static int seeded = 0;

    if (!seeded) {
        srand((unsigned int)time(NULL));
        seeded = 1;
    }
    for (int i = 0; i < 4; i++) {
        rolls[i] = rand() % 6 + 1;
    }
    qsort(rolls, 4, sizeof(int), compare);

    return rolls[1] + rolls[2] + rolls[3];
}

int modifier(int score) {
    return score / 2 - 5;
}

dnd_character_t make_dnd_character(void) {
    dnd_character_t character;

    character.strength     = ability();
    character.dexterity    = ability();
    character.constitution = ability();
    character.intelligence = ability();
    character.wisdom       = ability();
    character.charisma     = ability();
    character.hitpoints = 10 + modifier(character.constitution);

    return character;
}

File: solutions/lua/dnd-character/1/dnd-character.lua
local Character = {}

local function ability(scores)
    table.sort(scores)
    return scores[4] + scores[3] + scores[2]
end

local function roll_dice()
    return {
        math.random(6),
        math.random(6),
        math.random(6),
        math.random(6)
    }
end

local function modifier(input)
    return math.floor((input - 10) / 2)
end

function Character:new(name)
    character = {}
    character.name = name
    character.strength = ability(roll_dice())
    character.dexterity = ability(roll_dice())
    character.constitution = ability(roll_dice())
    character.intelligence = ability(roll_dice())
    character.wisdom = ability(roll_dice())
    character.charisma = ability(roll_dice())
    character.hitpoints = 10 + modifier(character.constitution)
    setmetatable(character, self)
    self.__index = self
    return character
end

return {
  Character = Character,
  ability = ability,
  roll_dice = roll_dice,
  modifier = modifier
}
